File: PODM/SW/pod-manager/podm-business-entities/src/main/java/com/intel/podm/business/entities/redfish/MemoryChunk.java
/*
 * Copyright (c) 2016 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.podm.business.entities.redfish;

import com.intel.podm.business.entities.base.DomainObject;
import com.intel.podm.business.entities.base.DomainObjectProperty;
import com.intel.podm.business.entities.redfish.base.Descriptable;
import com.intel.podm.business.entities.redfish.base.Discoverable;
import com.intel.podm.business.entities.redfish.base.StatusPossessor;
import com.intel.podm.common.types.AddressRangeType;
import com.intel.podm.common.types.Status;

import javax.enterprise.context.Dependent;
import javax.transaction.Transactional;
import java.net.URI;
import java.util.List;

import static com.intel.podm.business.entities.base.DomainObjectLink.CONTAINED_BY;
import static com.intel.podm.business.entities.base.DomainObjectLink.OWNED_BY;
import static com.intel.podm.business.entities.base.DomainObjectProperties.booleanProperty;
import static com.intel.podm.business.entities.base.DomainObjectProperties.enumProperty;
import static com.intel.podm.business.entities.base.DomainObjectProperties.integerProperty;
import static com.intel.podm.business.entities.base.DomainObjectProperties.stringProperty;
import static com.intel.podm.common.utils.IterableHelper.single;
import static com.intel.podm.common.utils.IterableHelper.singleOrNull;
import static javax.transaction.Transactional.TxType.REQUIRED;

@Dependent
@Transactional(REQUIRED)
public class MemoryChunk extends DomainObject implements Discoverable, StatusPossessor, Descriptable {
    public static final DomainObjectProperty<String> MEMORY_CHUNK_NAME = stringProperty("memoryChunkName");
    public static final DomainObjectProperty<Integer> MEMORY_CHUNK_UID = integerProperty("memoryChunkUid");
    public static final DomainObjectProperty<Integer> MEMORY_CHUNK_SIZE_MIB = integerProperty("memoryChunkSizeMib");
    public static final DomainObjectProperty<AddressRangeType> ADDRESS_RANGE_TYPE = enumProperty("addressRangeType", AddressRangeType.class);
    public static final DomainObjectProperty<Boolean> IS_MIRROR_ENABLED = booleanProperty("isMirrorEnabled");
    public static final DomainObjectProperty<Boolean> IS_SPARE = booleanProperty("isSpare");

    @Override
    public String getName() {
        return getProperty(NAME);
    }

    @Override
    public void setName(String name) {
        setProperty(NAME, name);
    }

    @Override
    public String getDescription() {
        return getProperty(DESCRIPTION);
    }

    @Override
    public void setDescription(String description) {
        setProperty(DESCRIPTION, description);
    }

    public String getMemoryChunkName() {
        return getProperty(MEMORY_CHUNK_NAME);
    }

    public void setMemoryChunkName(String memoryChunkName) {
        setProperty(MEMORY_CHUNK_NAME, memoryChunkName);
    }

    public Integer getMemoryChunkUid() {
        return getProperty(MEMORY_CHUNK_UID);
    }

    public void setMemoryChunkUid(Integer memoryChunkUid) {
        setProperty(MEMORY_CHUNK_UID, memoryChunkUid);
    }

    public Integer getMemoryChunkSizeMib() {
        return getProperty(MEMORY_CHUNK_SIZE_MIB);
    }

    public void setMemoryChunkSizeMib(Integer memoryChunkSizeMib) {
        setProperty(MEMORY_CHUNK_SIZE_MIB, memoryChunkSizeMib);
    }

    public AddressRangeType getAddressRangeType() {
        return getProperty(ADDRESS_RANGE_TYPE);
    }

    public void setAddressRangeType(AddressRangeType addressRangeType) {
        setProperty(ADDRESS_RANGE_TYPE, addressRangeType);
    }

    public Boolean getIsMirrorEnabled() {
        return getProperty(IS_MIRROR_ENABLED);
    }

    public void setIsMirrorEnabled(Boolean isMirrorEnabled) {
        setProperty(IS_MIRROR_ENABLED, isMirrorEnabled);
    }

    public Boolean getIsSpare() {
        return getProperty(IS_SPARE);
    }

    public void setIsSpare(Boolean isSpare) {
        setProperty(IS_SPARE, isSpare);
    }

    @Override
    public Status getStatus() {
        return getProperty(STATUS);
    }

    @Override
    public void setStatus(Status status) {
        setProperty(STATUS, status);
    }

    public List<Memory> getInterleaveSets() {
        return single(getLinked(CONTAINED_BY, ComputerSystem.class)).getMemoryModules();
    }

    @Override
    public URI getSourceUri() {
        return getProperty(SOURCE_URI);
    }

    @Override
    public void setSourceUri(URI sourceUri) {
        setProperty(SOURCE_URI, sourceUri);
    }

    @Override
    public ExternalService getService() {
        return singleOrNull(getLinked(OWNED_BY, ExternalService.class));
    }
}

File: PSME/common/agent-framework/src/module-ref/model/memory_chunk.cpp
/*!
 * @section LICENSE
 *
 * @copyright
 * Copyright (c) 2015-2016 Intel Corporation
 *
 * @copyright
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * @copyright
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * @copyright
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @section DESCRIPTION
 * @file memory_chunk.cpp
 * @brief Memory Chunk class
 *
 * */

#include "agent-framework/module-ref/model/memory_chunk.hpp"
#include "agent-framework/module-ref/constants/compute.hpp"
#include "agent-framework/module-ref/utils/utils.hpp"


using namespace agent_framework::model;
using namespace agent_framework::model::utils;

const enums::Component MemoryChunk::component = enums::Component::MemoryChunk;
const enums::CollectionName MemoryChunk::collection_name =
    enums::CollectionName::MemoryChunks;

MemoryChunk::MemoryChunk(const std::string& parent) : Resource{parent} {}

MemoryChunk::~MemoryChunk() {}

Json::Value MemoryChunk::to_json() const {
    Json::Value result;
    result[literals::MemoryChunk::STATUS] = get_status().to_json();
    result[literals::MemoryChunk::CHUNK_NAME] = get_chunk_name();
    result[literals::MemoryChunk::CHUNK_ID] = get_chunk_id();
    result[literals::MemoryChunk::CHUNK_SIZE] = get_chunk_size_mb();
    result[literals::MemoryChunk::TYPE] = get_type();
    result[literals::MemoryChunk::MIRRORED] = get_mirrored();
    result[literals::MemoryChunk::SPARE] = get_spare();
    result[literals::MemoryChunk::INTERLEAVE_SETS] = get_interleave_sets().to_json();
    result[literals::MemoryChunk::OEM] = get_oem().to_json();
    return result;
}

MemoryChunk MemoryChunk::from_json(const Json::Value& json) {
    MemoryChunk chunk;
    chunk.set_status(
        attribute::Status::from_json(json[literals::MemoryChunk::STATUS]));
    chunk.set_chunk_name(json[literals::MemoryChunk::CHUNK_NAME]);
    chunk.set_chunk_id(json[literals::MemoryChunk::CHUNK_ID]);
    chunk.set_chunk_size_mb(json[literals::MemoryChunk::CHUNK_SIZE]);
    chunk.set_type(json[literals::MemoryChunk::TYPE]);
    chunk.set_mirrored(json[literals::MemoryChunk::MIRRORED]);
    chunk.set_spare(json[literals::MemoryChunk::SPARE]);
    chunk.set_interleave_sets(
        InterleaveSets::from_json(json[literals::MemoryChunk::INTERLEAVE_SETS]));
    chunk.set_oem(attribute::Oem::from_json(json[literals::MemoryChunk::OEM]));
    return chunk;
}
